amazon-kinesis-client/src/main/java/software/amazon/kinesis/lifecycle/ShutdownTask.java

```diff
@@ -99,67 +104,48 @@ public TaskResult call() {
 
         try {
             try {
-                ShutdownReason localReason = reason;
-                List<Shard> latestShards = null;
-                /*
-                 * Revalidate if the current shard is closed before shutting down the shard consumer with reason SHARD_END
-                 * If current shard is not closed, shut down the shard consumer with reason LEASE_LOST that allows
-                 * active workers to contend for the lease of this shard.
-                 */
-                if (localReason == ShutdownReason.SHARD_END) {
-                    latestShards = shardDetector.listShards();
-
-                    //If latestShards is empty, should also shutdown the ShardConsumer without checkpoint with SHARD_END
-                    if (CollectionUtils.isNullOrEmpty(latestShards) || !isShardInContextParentOfAny(latestShards)) {
-                        localReason = ShutdownReason.LEASE_LOST;
-                        dropLease();
-                        log.info("Forcing the lease to be lost before shutting down the consumer for Shard: " + shardInfoIdProvider.apply(shardInfo));
+                log.debug("Invoking shutdown() for shard {}, concurrencyToken {}. Shutdown reason: {}",
+                          shardInfoIdProvider.apply(shardInfo), shardInfo.concurrencyToken(), reason);
+
+                final long startTime = System.currentTimeMillis();
+                if (reason == ShutdownReason.SHARD_END) {
+                    // Create new lease for the child shards if they don't exist.
+                    if (!CollectionUtils.isNullOrEmpty(childShards)) {
+                        createLeasesForChildShardsIfNotExist();
                     }
-                }
 
-                // If we reached end of the shard, set sequence number to SHARD_END.
-                if (localReason == ShutdownReason.SHARD_END) {
                     recordProcessorCheckpointer
                             .sequenceNumberAtShardEnd(recordProcessorCheckpointer.largestPermittedCheckpointValue());
                     recordProcessorCheckpointer.largestPermittedCheckpointValue(ExtendedSequenceNumber.SHARD_END);
-                }
-
-                log.debug("Invoking shutdown() for shard {}, concurrencyToken {}. Shutdown reason: {}",
-                        shardInfoIdProvider.apply(shardInfo), shardInfo.concurrencyToken(), localReason);
-                final ShutdownInput shutdownInput = ShutdownInput.builder().shutdownReason(localReason)
-                        .checkpointer(recordProcessorCheckpointer).build();
-                final long startTime = System.currentTimeMillis();
-                try {
-                    if (localReason == ShutdownReason.SHARD_END) {
+                    // Call the shardReocrdsProcessor to checkpoint with SHARD_END sequence number.
+                    // The shardEnded is implemented by customer. We should validate if the Shard_End checkpointing is successful after calling shardEnded.
```

Comment:
Let's simplify the exception handling as follows,
```
if(reason == ShutdownReason.SHARD_END) {
 ...
 throwOnApplicationException(() -> shardEndLogic, metricArgs)
} else {
 throwOnApplicationException(() -> leaseLostLogic, metricArgs)
}

private void throwOnApplicationException(Runnable action, MetricArgs..) throws ApplicationException {
try {
 action.run()
} catch (Exception e) {
 throw new ApplicationException(e);
} finally {
 emitMetrics(metricArgs)
}
```

Reply:
Addressed